### PR TITLE
00406-cve-2022-48565.patch

### PR DESCRIPTION
**Pull request by frenzymadness**
CVE-2022-48565: XML External Entity in XML processing plistlib module Backported from https://github.com/python/cpython/commit/a158fb9c5138db94adf24fbc5690467cda811163 Tracking bug: https://bugzilla.redhat.com/show_bug.cgi?id=CVE-2022-48565

### CODE REVIEW

**Comment by frenzymadness** [comment]
COPR builds: https://copr.fedorainfracloud.org/coprs/lbalhar/xml-CVE/builds/
Verified in tests and also by a reproducer.

**Comment by hroncok** [comment]
Could you please attribute this commit to the original authors? There is a link to the original commit, but the contribution seems to have been stripped.

In my head, the commit could look something like this:

```
From: Ned Deily <nad@python.org>
...
Subject: 00406-cve-2022-48565.patch

Reject XML entity declarations in plist files
CVE-2022-48565: XML External Entity in XML processing plistlib module
Backported from https://github.com/python/cpython/commit/a158fb9c5138
Tracking bug: https://bugzilla.redhat.com/show_bug.cgi?id=CVE-2022-48565

Co-authored-by: Ronald Oussoren <ronaldoussoren@mac.com>
Co-authored-by: Lumir Balhar <lbalhar@redhat.com>
```
-----


 - This patch is:  3 files changed, 34 insertions(+)
 - The original is: 3 files changed, 28 insertions(+)

I've checked the patches side by side. The patch file here mentions `\ No newline at end of file` for the news file, but those files are bit-by-bit identical, so this is not relevant.

The `parser.EntityDeclHandler = self.handle_entity_decl` addition only differs in `self` (which matches the surrounding code as well).

The extra lines appear to be the addition of `InvalidFileException`. I've checked that exception on Python 2 and it works as expected. Adding it makes sense.

The `EntityDeclHandler` thing apparently also exists on Python 2.7.

----

The differences in the test were unexpected to me.

Here there is:

```python
    def test_xml_plist_with_entity_decl(self):
        with self.assertRaises(plistlib.InvalidFileException) as context:
            plistlib.readPlistFromString(XML_PLIST_WITH_ENTITY)

        self.assertTrue("XML entity declarations are not supported" in str(context.exception))
```

But the original had:

```python
    def test_xml_plist_with_entity_decl(self):
        with self.assertRaisesRegex(plistlib.InvalidFileException,
                                    "XML entity declarations are not supported"):
            plistlib.loads(XML_PLIST_WITH_ENTITY, fmt=plistlib.FMT_XML)
```

 - I assume the `.loads` method is unavailable on Python 2.7. At least I don't see it in the docs. Using a different method name makes sense to me.
 - What was the reason to restructure the test to use `context`? Is [assertRaisesRegexp](https://docs.python.org/2.7/library/unittest.html#unittest.TestCase.assertRaisesRegexp) not sufficient? (The different way to assert the message confused me.)

Nevertheless, the test seems to assert the same thing. If the tests runs and passes, I'd consider this fix OK.

File: Lib/test/test_plistlib.py [hunk 1]
```diff
@@ -86,6 +86,19 @@
 </plist>
 """.replace(" " * 8, "\t")  # Apple as well as plistlib.py output hard tabs
 
+XML_PLIST_WITH_ENTITY = b'''\
```
Comment (hroncok) [inline Lib/test/test_plistlib.py]: If you keep this ugly (as original), it's easier to compare with the backported commit and hence easier to review.


```suggestion
XML_PLIST_WITH_ENTITY=b'''\
```

**Comment by frenzymadness** [comment]
Commit author adjusted as you wish.

>     * What was the reason to restructure the test to use `context`? Is [assertRaisesRegexp](https://docs.python.org/2.7/library/unittest.html#unittest.TestCase.assertRaisesRegexp) not sufficient? (The different way to assert the message confused me.)

I had to change that, the previous version failed with `AttributeError: 'TestPlistlib' object has no attribute 'assertRaisesRegex'`, see the middle build in the linked COPR.

File: Lib/test/test_plistlib.py (hunk 1 above)
Comment (frenzymadness) [reply Lib/test/test_plistlib.py]: Improvement reverted.

**Comment by hroncok** [comment]
It was called [assertRaisesRegexp](https://docs.python.org/2.7/library/unittest.html#unittest.TestCase.assertRaisesRegexp) on Python 2.

**Comment by frenzymadness** [comment]
Fixed

**Review by hroncok — APPROVED** [review]

**Pull request merged**